Fixes 5642 for Go: allow DependsOn accepting inputs and outputs

t0yv0 opened the pull request:
# Description

<!--- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. -->

Fixes #5642 (for the Go language)

## Checklist

<!--- Please provide details if the checkbox below is to be left unchecked. -->
- [ ] I have added tests that prove my fix is effective or that my feature works
<!--- 
User-facing changes require a CHANGELOG entry.
-->
- [ ] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change
<!--
If the change(s) in this PR is a modification of an existing call to the Pulumi Service,
then the service should honor older versions of the CLI where this change would not exist.
You must then bump the API version in /pkg/backend/httpstate/client/api.go, as well as add
it to the service.
-->
- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version
  <!-- @Pulumi employees: If yes, you must submit corresponding changes in the service repo. -->

t0yv0 commented on sdk/go/pulumi/resource.go:
```diff
@@ -314,7 +316,69 @@ func DeleteBeforeReplace(o bool) ResourceOption {
+// A fully dynamic version of `DependsOn` that accepts `AnyOutput`
+// carrying `[]Resource`, `[]ResourceOutput`, or `[]ResourceInput`
+// values, which are then considered as explicit dependencies.
+func DependsOnOutput(any AnyOutput) ResourceOption {
```
Including to be compatible in power with Node/Py SDK where a list of prompty-indeterminate length can be passed in. If we don't care about that, can remove. It's a bit of an unfortunate interface.

t0yv0 commented on sdk/go/pulumi/resource.go:
```diff
@@ -314,7 +316,69 @@ func DeleteBeforeReplace(o bool) ResourceOption {
+		ro.DependsOn = append(ro.DependsOn, func(ctx context.Context) ([]URN, error) {
+			var urns []URN
+			for _, res := range o {
+				directUrn, isKnown, _, err := res.URN().awaitURN(ctx)
```
Any dependencies present in the `Output` layer of `res.URN() : URNOutput` are ignored.

pgavlin replied:
any particular reason for this? does this match the behavior of the other SDKs?

t0yv0 replied:
I can test for it. Node can be our reference. I'd need a way to inject deps into the URN output but that perhaps is possible.

t0yv0 replied:
Matches the behavior of Node. The dependencies specifically on URNOutput do not participate. C# SDK diverges significantly https://github.com/pulumi/pulumi/pull/7611 and I'm taking Node as authority on what should happen.

t0yv0 commented on sdk/go/pulumi/resource.go:
```diff
@@ -433,3 +497,92 @@ func Version(o string) ResourceOrInvokeOption {
+}
+
+// Converts boxed `[]Resource`, `[]ResourceOutput`, `[]ResourceInput` slices to `[]ResourceInput`.
+func autocastSliceToResourceInput(v interface{}) ([]ResourceInput, error) {
```
In Go, slices seem to be invariant, one can cast `CustomResource -> interface{} -> Resource` but one cannot cast `[]CustomResource -> interface{} -> []Resource`... Rolling a little helper here. If there is other machinery in the codebase that can do this for our users please LMK.

t0yv0 commented on sdk/go/pulumi/resource_test.go:
```diff
@@ -317,3 +348,231 @@ func TestResourceOptionMergingReplaceOnChanges(t *testing.T) {
+	return &dependenciesTracker{dependsOn}
+}
+
+type interceptingResourceMonitor struct {
```
Brute force attack here on the problem of intercepting engine calls to reconstruct dependency map "as reported to the enigne". Please LMK if there are other more elegant ways. Perhaps build this into mockEngine instead of mockMonitor?

pgavlin replied:
Ah, is this necessary b/c we don't include dependency info in `MockResourceArgs`?

t0yv0 replied:
Right, the other way would be to package this into MockResourceArgs, simpler implementation. That'd be a public interface change and up for consideration if we want that public change or not, I did not want to tag that along here. OK to tackle separately or you'd prefer to go ahead with extending MockResourceArgs and I can make this PR depend on it?

t0yv0 commented on sdk/go/pulumi/types.go:
```diff
@@ -1017,6 +1017,31 @@ func (ResourceOutput) ElementType() reflect.Type {
+
+// An Input type carrying Resource values.
+//
+// Unfortunately `Resource` values do not implement `ResourceInput` in
```
I'm running into a lot of difficulty trying to make `Resource` values compatible with `ResourceInput`. I'm starting to think the problem is with "deep reflect" combinators that try to traverse values as trees, find all Inputs and await them.. 

I also worry about these combinators attempting to await extraneous/outputs inputs on Resource properties and try to count them toward dependencies. If we have `ResourceInput` carrying a `r: Resource` , when we count the dependency closure, we want dependencies of the `Input[]` and the `r.urn`, and we probably don't want the deps of `r.Output1` etc.

pgavlin replied:
> I'm running into a lot of difficulty trying to make Resource values compatible with ResourceInput. I'm starting to think the problem is with "deep reflect" combinators that try to traverse values as trees, find all Inputs and await them..

Can you elaborate on the issues you're running into?

> we probably don't want the deps of r.Output1 etc.

FWIW, in practice the deps of `r.Output1` for a custom resource are `[]Resource{r}`. But I think that you're right in general.

t0yv0 replied:
Very specifically this: https://github.com/pulumi/pulumi/pull/7394

Substitute `Example` with `Resource` to make this practical. When interfaces travel through input/output machinery something in the generic/reflecty code gets off the rails, and I tried but failed to fix that.

justinvp replied:
With the change in #7394, are you suggesting we'd change `Resource` to implement `ResourceInput`?  Or would we make `ResourceState` implement `ResourceInput`?

Is that change necessary for this PR or can it be deferred?

t0yv0 replied:
I'm struggling to make the suggestion more specific. For the users' sake, I felt it might be "nice" to be able to pass resources where resource inputs are expected, e.g.:

```
   func f(x ResourceInput)
   res := NewFooResource("my-res1")
   f(res)

   # vs
   
   f(NewResourceInput(res))
```

Initially I thought this would need:

```
   type Resource interface {
        ...
	ResourceInput
   }
```

But your comment makes me realize specific resource types can implement ResourceInput without making Resource implement ResourceInput. So far so good.

It was tempting to make some changes to `ResourceState` so that *all* structs embedding ResourceState automatically start implementing ResourceInput. This however is not possible AFAIK, because:

```
type Input interface {
	ElementType() reflect.Type
}
type ResourceInput interface {
        Input
}
```

So `ResourceState` cannot implement `ElementType()` since it does not know the struct it is being embedded in, and AFAIK it cannot capture `self` referring to that outer struct. This is somewhat unlike classical inheritance.

t0yv0 replied:
Now that I think of it, we could follow up with a codegen change though to make provider resources implement ResourceInput.

justinvp replied:
These are often going to show up as outputs from multi-lang components like [this](https://github.com/pulumi/pulumi-eks/blob/86861e6cb6091e97d08828a444bb70087a8701f3/sdk/go/eks/managedNodeGroup.go#L24):

```go
type ManagedNodeGroup struct {
	pulumi.ResourceState

	// The AWS managed node group.
	NodeGroup eks.NodeGroupOutput `pulumi:"nodeGroup"`
}
```

If I have an instance of the `ManagedNodeGroup` component, and I want to pass its `NodeGroup` output (typed as `github.com/pulumi/pulumi-aws/sdk/v4/go/aws/eks.NodeGroupOutput`) to `DependsOnInputs`, what would that look like with just the changes in this PR?

Would I have to use an `ApplyT` to do the conversion like the following?

```go
ng, err := eks.NewManagedNodeGroup(ctx, "ng", &eks.ManagedNodeGroupArgs{...})
pulumi.DependsOnInputs(pulumi.ResourceArray{
	ng.NodeGroup.ApplyT(func(v pulumi.Resource) pulumi.Resource {
		return v
	}).(pulumi.ResourceOutput),
})
```

t0yv0 replied:
I filed this as https://github.com/pulumi/pulumi/issues/7750

t0yv0 commented:
> Should we add a ResourceArray/ResourceArrayInput/ResourceArrayOutput and then just have a single function: func DependsOnInputs(o ResourceArrayInput) ResourceOption?

That sounds interesting, it can lead to a better interface, I've just been a little unsure how to implement input types completely correctly in the Go SDK (I see you've commented 🤔  on some linked PRs). Let me dig a little deeper on this, I feel like I need to put an effort here to understand this matter better.

t0yv0 commented on sdk/go/pulumi/provider.go:
```diff
@@ -103,7 +103,11 @@ func construct(ctx context.Context, req *pulumirpc.ConstructRequest, engineConn
 	opts := resourceOption(func(ro *resourceOptions) {
 		ro.Aliases = aliases
-		ro.DependsOn = dependencies
+		ro.DependsOn = []func(ctx context.Context) ([]URN, error){
```
Oh no, this is gofmt. In my Emacs mode it runs gofmt on every file save. If I add a space it removes it.

t0yv0 commented on sdk/go/pulumi/types.go:
```diff
@@ -1017,6 +1017,116 @@ func (ResourceOutput) ElementType() reflect.Type {
+	return NewResourceArray(in...)
+}
+
+func NewResourceArray(in ...Resource) ResourceArray {
```
The "standard" code generator  builds the ToResourceArray forms, but I find the `...` forms much more handy to use. I included both so as to not make inconsistent To* forms.

t0yv0 commented:
Added `func DependsOnInputs(o ResourceArrayInput) ResourceOption` - this works pretty well. We lose parity vs Python in a niche matter of unknowns, that is, since casting to ResourceArrayInput uses `All` if any of the resources are unknown, then no dependencies at all will be recorded. Whereas in Python, if any of them are known they are still recorded. I've no evidence it's a practical issue, just a theoretical curiosity, so suggest we ignore. I removed the test that was looking into that.

In current version `NewResourceInput` is still needed. To make `Resource` values automatically promote to `ResourceInput` we need two things:

- Fix core lib handling of outputs carrying interfaces https://github.com/pulumi/pulumi/pull/7394

- Embed `ResourceInput` interface into `Resource` interface; it seems that it would that break providers and might necessitate a codegen change, as I don't believe we can patch ResourceState and make all resources implement ResourceInput correctly; this will also be a breaking change to user code for custom resources as they will not be implementing ResourceInput automatically

Given these two considerations, I suggest we postpone that and accept the current version necessitating NewResourceInput for now?

t0yv0 commented:
@justinvp PTAL

justinvp reviewed (approved):
One question about #7394, otherwise LGTM, aside from any other feedback from @pgavlin or @EvanBoyle.

t0yv0 commented on sdk/go/pulumi/urnset.go:
```diff
@@ -0,0 +1,50 @@
+	return values
+}
+
+func (s urnSet) sortedValues() []URN {
```
@pgavlin as a compromise I just added this to use in my tests, not making values() sort.
